### ISSUE TITLE
[flytekit/async] Context manager & general thread/coroutine safety

### ISSUE THREAD

**Issue by wild-endeavor** [flyteorg/flyte#6073]
## Overview
The flytekit sdk has been traditionally developed without much regard for thread-safety and simultaneous access of objects.  With the introduction of async support and the ability to run multiple tasks concurrently if not in parallel, we should revisit these assumptions, fix and test bugs and limitations.

This is also likely going to be a problem with the new [async workflows](https://github.com/flyteorg/flyte/issues/6051) effort that will be done.

The context manager is a known issue but we should find others as well.

### Context Manager
The context manager probably is the biggest offender here.  This is a global list of context states and is appended to every time a new task is run, among other operations.  This breaks in the new eager model for instance.  If you have the following series of tasks.

```python
@task
def t1():
    print("normal task")

@eager
def eg_1():
    some_other_task()

@eager
def eg_p():
    t1()
    eg_1()
```

because each of `eg_p`, `t1`, and `eg_1` will try to set context by appending something to the list, the view of what is `FlyteContextManager.current_context()` will not make sense.

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by kumare3** [comment]
We are closing a bunch of issues, are we are getting ready to update a V2
